sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala

```diff
@@ -315,9 +335,17 @@ private[hive] object HadoopTableReader extends HiveInspectors {
       }
     }
 
+    /**
+     * when the soi and deserializer.getObjectInspector is equal,
+     * we will get `IdentityConverter`,which mean it won't convert the
+     * value when schema match
+     */
+    val partTblObjectInspectorConverter = ObjectInspectorConverters.getConverter(
+      deserializer.getObjectInspector, soi)
+
     // Map each tuple to a row object
     iterator.map { value =>
-      val raw = deserializer.deserialize(value)
+      val raw = partTblObjectInspectorConverter.convert(deserializer.deserialize(value))
```

Comment:
There logic here is a little confusing for me.
As we already have the `converter` here, probably we don't need to call the `getConvertedOI`, and the `soi` should be the expected output `ObjectInspector`, which supposed to be the output object inspector of from the table deserializer. right?